### DISCUSSION TITLE
[FEATURE] Stubs for python language server

### DISCUSSION THREAD

**Discussion by John-Schreiber** [pyscript/pyscript#1173] [Proposals]
It would be useful to have stubs for the pyodide and js modules so syntax highlighting and autocorrect work. Pyodide has https://github.com/pyodide/pyodide/issues/1839, but I can't get it to work.

**Comment by marimeireles** [comment]
It would!
Interesting thing to tackle. :)

**Reply by microhobby** [reply]
oh no, since 2022. 
It would be great to have these stubs to use VS Code for example with the completions and mypy checks.

**Reply by microhobby** [reply]
@John-Schreiber the pyodide, as I understand, is the runtime. We will need the stubs for the pyscript module. Also, the `document` and `window` for example, and maybe others, are runtime bind. So, it only exists inside the wasm runtime. Maybe I'm talking shit, but it is what I understand.

**Reply by WebReflection** [reply]
`document` and `window` are the entry point to the entirety of the JS / Web world and I've no idea how to tell the IDE to switch code hints or completion once those two things are referenced ... the `pyscript` module itself has many things to hint/document but from `window` or `document` ref on, it's a Web world, not Python anymore ... yet I'd love to have something first, so then we can thing about the "*something else too*"

**Reply by WebReflection** [reply]
P.S. on a second though ... how is Pyodide doing with its `js` import? if they found a solution there, we might as well reuse the exact same, as `from pyodide import js` is our equivalent of `from pyscript import window` (except our case is about the main thread world, not the current worker scope, when in workers)

**Reply by Josverl** [reply]
There is a webassembly version of the MicroPython-stubs on PyPI.
You can `pip install micropython-webassembly-stubs --target typings`  into a local folder and configure you IDE / typechecker / CI to use that.

its not perfect - if you want to help make it better then please take a look at improving [the reference stubs](https://github.com/Josverl/micropython-stubs/tree/main/reference/pyscript) that are used to build the actual stub packages.

Improving types , docstrings can be without much effort.

Also let me know if you want to help, but find my documentation on how to do that too terse. Then I know where to improve.
